server/interceptor/interceptor.go

```diff
@@ -29,43 +29,52 @@ import (
 	"google.golang.org/grpc/status"
 )
 
-// TreeInterceptor ensures that all requests pertaining a specific tree are of correct type and
-// respect the current tree state (frozen, deleted, etc).
-type TreeInterceptor struct {
-	Admin storage.AdminStorage
+// TrillianInterceptor checks that:
+// * Requests addressing a tree have the correct tree type and tree state;
+// * Requests are properly authenticated / authorized (NOT YET DONE); and
+// * Requests are rate limited appropriately.
+type TrillianInterceptor struct {
+	Admin        storage.AdminStorage
+	QuotaManager quota.Manager
 }
 
-// UnaryInterceptor executes the TreeInterceptor logic for unary RPCs.
-func (i *TreeInterceptor) UnaryInterceptor(ctx context.Context, req interface{}, info *grpc.UnaryServerInfo, handler grpc.UnaryHandler) (interface{}, error) {
-	rpcInfo, err := getRPCInfo(req, info.FullMethod)
+// UnaryInterceptor executes the TrillianInterceptor logic for unary RPCs.
+func (i *TrillianInterceptor) UnaryInterceptor(ctx context.Context, req interface{}, info *grpc.UnaryServerInfo, handler grpc.UnaryHandler) (interface{}, error) {
+	quotaUser := i.QuotaManager.GetUser(ctx, req)
+
+	rpcInfo, err := getRPCInfo(req, info.FullMethod, quotaUser)
 	if err != nil {
 		return nil, err
 	}
 
-	if rpcInfo.doesNotHaveTree {
-		return handler(ctx, req)
+	if rpcInfo.treeID != 0 {
+		tree, err := trees.GetTree(ctx, i.Admin, rpcInfo.treeID, rpcInfo.opts)
+		if err != nil {
+			return nil, err
+		}
+		ctx = trees.NewContext(ctx, tree)
+
+		// TODO(codingllama): Add auth interception
 	}
 
-	tree, err := trees.GetTree(ctx, i.Admin, rpcInfo.treeID, rpcInfo.opts)
-	if err != nil {
-		return nil, err
+	if err := i.QuotaManager.GetTokens(ctx, 1 /* numTokens */, rpcInfo.specs); err != nil {
+		return nil, status.Errorf(codes.ResourceExhausted, "quota exhausted: %v", err)
 	}
 
-	ctx = trees.NewContext(ctx, tree)
 	return handler(ctx, req)
 }
 
 // rpcInfo contains information about an RPC, as extracted from its request message.
 type rpcInfo struct {
-	// doesNotHaveTree states whether the RPC is tied to a specific tree.
-	// Examples of RPCs without trees are CreateTree (no tree exists yet) and ListTrees
-	// (zero to many trees returned).
-	doesNotHaveTree bool
-	// treeID is the tree ID tied to this RPC, if any.
+	// treeID is the tree ID tied to this RPC, if any (zero means no tree).
```

Comment:
Not worth changing now, but this change would have been better as a distinct commit.